lte/gateway/python/integ_tests/s1aptests/test_attach_detach_ipv6.py

```diff
@@ -0,0 +1,84 @@
+"""
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import time
+import unittest
+
+import s1ap_types
+from integ_tests.s1aptests import s1ap_wrapper
+
+
+class TestAttachDetachIpv6(unittest.TestCase):
+
+    def setUp(self):
+        self._s1ap_wrapper = s1ap_wrapper.TestWrapper()
+
+    def tearDown(self):
+        self._s1ap_wrapper.cleanup()
+
+    def test_attach_detach_ipv6(self):
+        """ Basic attach/detach IPv6 test with single UE """
+        num_ues = 2
+        detach_type = [
+            s1ap_types.ueDetachType_t.UE_NORMAL_DETACH.value,
+            s1ap_types.ueDetachType_t.UE_SWITCHOFF_DETACH.value,
+        ]
+        wait_for_s1 = [True, False]
+        self._s1ap_wrapper.configUEDevice(num_ues)
+
+        # Default apn over-write
+        magma_default_apn = {
+            "apn_name": "magma.ipv4",  # APN-name
+            "qci": 9,  # qci
+            "priority": 15,  # priority
+            "pre_cap": 1,  # preemption-capability
+            "pre_vul": 0,  # preemption-vulnerability
+            "mbr_ul": 200000000,  # MBR UL
+            "mbr_dl": 100000000,  # MBR DL
+            "pdn_type": 2,  # PDN Type 0-IPv4,1-IPv6,2-IPv4v6
+        }
+
+        apn_list = [magma_default_apn]
+
+        for i in range(num_ues):
+            req = self._s1ap_wrapper.ue_req
+            print(
+                "************************* Running End to End attach for ",
+                "UE id ", req.ue_id,
+            )
+            self._s1ap_wrapper.configAPN(
+                "IMSI" + "".join([str(j) for j in req.imsi]), 
+                apn_list, default=False
+            )
+            # Now actually complete the attach
+            self._s1ap_wrapper._s1_util.attach(
+                req.ue_id, s1ap_types.tfwCmd.UE_END_TO_END_ATTACH_REQUEST,
+                s1ap_types.tfwCmd.UE_ATTACH_ACCEPT_IND,
+                s1ap_types.ueAttachAccept_t,
+                pdn_type=2,
+            )
+
+            # Wait on EMM Information from MME
+            self._s1ap_wrapper._s1_util.receive_emm_info()
+            print(
+                "************************* Running UE detach for UE id ",
+                req.ue_id,
+            )
+            # Now detach the UE
```

Comment:
can we confirm if UE has IPv6 address same basic sanity test.

Reply:
Tester prints out the IP address and solicitation message. Since local interface address is randomly assigned by mobilityd, not sure what your ask is.

Reply:
this can be just type check if UE IP is valid IPv6 address in IP-POOL range of mobilityD configured pool

Reply:
@pshelar I don't think s1ap tester itself has a utility for this. This also looks more relevant for unit test coverage for mobilityd, isn't it? Whether the info is correctly encoded or not should be covered by the unit tests on MME.

@VinashakAnkitAman @pruthvihebbani, does S1AP tester validate received address types. Also is there a mechanism where we can pass expected values to S1AP tester out of band generically and assert on them?

Reply:
I guess there is mobilityD test for IPv6 allocator.
Its good to have UE IPv6 address validation in integration test for E2E coverage.

Reply:
@ulaskozat In the testcase we can validate the PDN type received in attach accept. e.g : self.assertEqual(attach.esmInfo.pAddr.pdnType, 3/*for ipv4v6*/). However s1ap tester cannot validate the ip address assigned by nw. One thing that can be done is if we already know the range within which the ip will be allocated from mobilityd, then we can configure this ip range during configUEDevice() and after receiving the ip address in attach accept s1ap tester can check if the ip is within this range

Reply:
ok, lets target it as followup.